File: Basic/Test_ReplaceElementswithGreatestElementonRightSide.c
// https://leetcode.com/problems/replace-elements-with-greatest-element-on-right-side/

#include <stdio.h>

int max(int num1, int num2)
{
    int result;
    if (num1 > num2)
    {
        result = num1;
    }
    else
    {
        result = num2;
    }
    return result;
}

void Solution(int *arr, int aSize)
{
    int maxSofar = arr[aSize - 1];
    arr[aSize - 1] = -1;
    for (int i = aSize - 2; i >= 0; i--)
    {
        int temp = arr[i];
        arr[i] = maxSofar;
        maxSofar = max(temp, maxSofar);
    }
}

int main()
{
    int t1[6] = {17, 18, 5, 4, 6, 1};
    Solution(t1, 6);
    for (int i = 0; i < 6; i++)
    {
        printf("%d ", t1[i]);
    }
    return 0;
}

File: Basic/Test_ReplaceElementswithGreatestElementonRightSide.py
# https://leetcode.com/problems/replace-elements-with-greatest-element-on-right-side/

def Solution(arr=list):
    maxSoFar = arr[len(arr) - 1]
    arr[len(arr) - 1] = -1
    # for backward
    for i in range(len(arr) - 2, -1, -1):
        temp = arr[i]
        arr[i] = maxSoFar
        maxSoFar = max(maxSoFar, temp)
    return arr


print(Solution(arr=[17, 18, 5, 4, 6, 1]))
